src/commcare_cloud/environment/secrets/backends/ansible_vault/main.py

```diff
@@ -0,0 +1,147 @@
+import getpass
+import os
+import sys
+from contextlib import contextmanager
+
+import datadog.api
+import yaml
+from ansible.parsing.vault import AnsibleVaultError
+from ansible_vault import Vault
+from memoized import memoized
+from six.moves import shlex_quote
+
+from commcare_cloud.environment.paths import ANSIBLE_DIR
+from commcare_cloud.environment.secrets.secrets_schema import get_known_secret_specs
+
+
+class AnsibleVaultSecretsBackend(object):
+    def __init__(self, name, vault_file_path, record_to_datadog=False):
+        self.name = name
+        self.vault_file_path = vault_file_path
+        self.record_to_datadog = record_to_datadog
+        self.should_send_vault_loaded_event = True
+
+    def prompt_user_input(self):
+        # call this for its side-effect: asking the user for the vault password
+        # (and thus not requiring that thereafter)
+        self._get_ansible_vault_password_and_record()
+
+    @staticmethod
+    def get_extra_ansible_args():
+        return (
+            '--vault-password-file={}/echo_vault_password.sh'.format(ANSIBLE_DIR),
+        )
+
+    def get_extra_ansible_env_vars(self):
+        return {
+            'ANSIBLE_VAULT_PASSWORD': self._get_ansible_vault_password_and_record(),
+        }
+
+    @staticmethod
+    def get_generated_variables():
+        secret_specs = get_known_secret_specs()
+        secret_specs_by_name = {secret_spec.name: secret_spec for secret_spec in secret_specs}
+        generated_variables = {}
+        for secret_spec in secret_specs:
```

Comment:
Hmm that's interesting. My first instinct was that this is going to be implemented differently depending on the secrets backend; this function is a core thing that'll be different between secrets backends because it's about telling ansible how to fetch each of the secrets and assign it to a variable. But on second thought, I think all the expression building (e.g. `| default(None)`) after the basic fetch syntax (e.g. `localsettings_private.SECRET_KEY`) might be pretty fixed regardless of the implementation. I'll think about what the common denominator is when I'm putting together the first other backend.